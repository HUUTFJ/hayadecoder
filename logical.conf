synchronous_commit = 'on'
wal_level = logical
max_replication_slots = 1