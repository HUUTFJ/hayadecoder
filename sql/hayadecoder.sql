SELECT slot_name FROM pg_create_logical_replication_slot('custom_slot', 'hayadecoder');

SELECT slot_name, plugin, slot_type FROM pg_replication_slots WHERE slot_name = 'custom_slot';

CREATE TABLE foo (id int);

BEGIN;
INSERT INTO foo VALUES (1);
UPDATE foo SET id = 2 WHERE id = 1;
DELETE FROM foo;
COMMIT;

SELECT data FROM pg_logical_slot_peek_changes('custom_slot', NULL, NULL, 'write_txd', 'off');
