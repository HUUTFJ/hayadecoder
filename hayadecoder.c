#include "postgres.h"

#include "catalog/pg_class.h"

#include "replication/logical.h"
#include "replication/origin.h"
#include "utils/builtins.h"
#include "utils/relcache.h"

extern void _PG_init(void);
extern void _PG_output_plugin_init(OutputPluginCallbacks *cb);

PG_MODULE_MAGIC;

typedef struct hayadecoder_data
{
    bool write_txd;
} hayadecodeData;

/* support routines */
static void output_insert(StringInfo out,
                          Relation relation,
			  ReorderBufferChange *change);
static void output_update(StringInfo out,
                          Relation relation,
			  ReorderBufferChange *change);
static void output_delete(StringInfo out,
                          Relation relation,
			  ReorderBufferChange *change);

/* callback routines */
static void hayade_startup(LogicalDecodingContext *ctx,
                           OutputPluginOptions *options,
                           bool is_init);
static void hayade_shutdown(LogicalDecodingContext *ctx);
static void hayade_begin(LogicalDecodingContext *ctx, ReorderBufferTXN *txn);
static void hayade_change(LogicalDecodingContext *ctx,
                      ReorderBufferTXN *txn,
                      Relation relation,
                      ReorderBufferChange *change);
static void hayade_commit(LogicalDecodingContext *ctx,
                      ReorderBufferTXN *txn,
                      XLogRecPtr commit_lsn);

static void
hayade_startup(LogicalDecodingContext *ctx,
                           OutputPluginOptions *options,
                           bool is_init)
{
    hayadecodeData *data;
    ListCell *option;

    data = palloc(sizeof(hayadecodeData));
    data->write_txd = true;

    options->output_type = OUTPUT_PLUGIN_TEXTUAL_OUTPUT;
    foreach(option, ctx->output_plugin_options)
    {
        DefElem *elem = lfirst(option);

        if (strcmp(elem->defname, "write_txd") == 0)
        {
            if (elem->arg == NULL)
                data->write_txd = true;
            else if (!parse_bool(strVal(elem->arg), &data->write_txd))
                elog(ERROR, "write_txd: wrong style");
        }
        else
            elog(ERROR, "unknown option");
    }

    ctx->output_plugin_private = data;
}

static void
hayade_shutdown(LogicalDecodingContext *ctx)
{
    hayadecodeData *data = ctx->output_plugin_private;

    pfree(data);
}

static void
hayade_begin(LogicalDecodingContext *ctx, ReorderBufferTXN *txn)
{
    hayadecodeData *data = ctx->output_plugin_private;

    OutputPluginPrepareWrite(ctx, true);

    if (data->write_txd)
        appendStringInfo(ctx->out, "START TRANSACTION %u!", txn->xid);
	else
        appendStringInfo(ctx->out, "START TRANSACTION!");

    OutputPluginWrite(ctx, true);
}

static void
hayade_change(LogicalDecodingContext *ctx,
                      ReorderBufferTXN *txn,
                      Relation relation,
                      ReorderBufferChange *change)
{
    switch (change->action)
    {
        case REORDER_BUFFER_CHANGE_INSERT:
            OutputPluginPrepareWrite(ctx, true);
            output_insert(ctx->out, relation, change);
            OutputPluginWrite(ctx, true);
            break;
        case REORDER_BUFFER_CHANGE_UPDATE:
            OutputPluginPrepareWrite(ctx, true);
            output_update(ctx->out, relation, change);
            OutputPluginWrite(ctx, true);
            break;
        case REORDER_BUFFER_CHANGE_DELETE:
            OutputPluginPrepareWrite(ctx, true);
            output_delete(ctx->out, relation, change);
            OutputPluginWrite(ctx, true);
            break;
        default:
            elog(ERROR, "unknown change");
            break;
    }
}

static void
output_insert(StringInfo out,
	      Relation relation,
	      ReorderBufferChange *change)
{
    Form_pg_class rel;
    char *relname;

    Assert(change->action == REORDER_BUFFER_CHANGE_INSERT);

    rel = relation->rd_rel;
    relname = NameStr(rel->relname);

    appendStringInfo(out, "INSERTED ");
    appendStringInfo(out, "TABLE: %s ", relname);
}

static void
output_update(StringInfo out,
	      Relation relation,
	      ReorderBufferChange *change)
{
    Form_pg_class rel;
    char *relname;

    Assert(change->action == REORDER_BUFFER_CHANGE_UPDATE);

    rel = relation->rd_rel;
    relname = NameStr(rel->relname);

    appendStringInfo(out, "UPDATED");
    appendStringInfo(out, "TABLE: %s ", relname);
}

static void
output_delete(StringInfo out,
	      Relation relation,
	      ReorderBufferChange *change)
{
    Form_pg_class rel;
    char *relname;

    Assert(change->action == REORDER_BUFFER_CHANGE_DELETE);

    rel = relation->rd_rel;
    relname = NameStr(rel->relname);

    appendStringInfo(out, "DELETED");
    appendStringInfo(out, "TABLE: %s ", relname);
}

static void
hayade_commit(LogicalDecodingContext *ctx,
                      ReorderBufferTXN *txn,
                      XLogRecPtr commit_lsn)
{
    hayadecodeData *data = ctx->output_plugin_private;

    OutputPluginPrepareWrite(ctx, true);

    if (data->write_txd)
        appendStringInfo(ctx->out, "END OF TRANSACTION %u!", txn->xid);
	else
        appendStringInfo(ctx->out, "END OF TRANSACTION!");

    OutputPluginWrite(ctx, true);
}

/* Public Routines */

void
_PG_init(void)
{
    /* NOP */
}

void
_PG_output_plugin_init(OutputPluginCallbacks *cb)
{
	AssertVariableIsOfType(&_PG_output_plugin_init, LogicalOutputPluginInit);

    cb->startup_cb = hayade_startup;
    cb->shutdown_cb = hayade_shutdown;
    cb->begin_cb = hayade_begin;
    cb->change_cb = hayade_change;
    cb->commit_cb = hayade_commit;
}
