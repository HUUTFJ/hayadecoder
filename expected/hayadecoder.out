SELECT slot_name FROM pg_create_logical_replication_slot('custom_slot', 'hayadecoder');
  slot_name  
-------------
 custom_slot
(1 row)

SELECT slot_name, plugin, slot_type FROM pg_replication_slots WHERE slot_name = 'custom_slot';
  slot_name  |   plugin    | slot_type 
-------------+-------------+-----------
 custom_slot | hayadecoder | logical
(1 row)

CREATE TABLE foo (id int);
BEGIN;
INSERT INTO foo VALUES (1);
UPDATE foo SET id = 2 WHERE id = 1;
DELETE FROM foo;
COMMIT;
SELECT data FROM pg_logical_slot_peek_changes('custom_slot', NULL, NULL, 'write_txd', 'off');
         data         
----------------------
 START TRANSACTION!
 END OF TRANSACTION!
 START TRANSACTION!
 INSERTED TABLE: foo 
 UPDATEDTABLE: foo 
 DELETEDTABLE: foo 
 END OF TRANSACTION!
(7 rows)

